apiserver/internal/handlers/resource/validate/validate.go

```diff
@@ -0,0 +1,211 @@
+
+	// ErrUnexpectedSize is returned when the size of the downloaded resources
+	// does not match the expected size.
+	ErrUnexpectedSize = errors.ConstError("downloaded resource has unexpected size (this could be because the controller disk does not have enough space)")
+)
+
+// FileSystem defines a file system for modifying files on a users system.
+type FileSystem interface {
+	// CreateTemp creates a new temporary file in the directory dir,
+	// opens the file for reading and writing, and returns the resulting file.
+	CreateTemp(dir, pattern string) (*os.File, error)
+
+	// Open opens the named file for reading. If successful, methods on
+	// the returned file can be used for reading; the associated file
+	Open(name string) (*os.File, error)
+
+	// Remove removes the named file or (empty) directory.
+	Remove(name string) error
+
+	// Statfs returns information about the file system.
+	Statfs(path string, stat *syscall.Statfs_t) error
+}
+
+// Validator validates a resource blob.
+type Validator struct {
+	logger     logger.Logger
+	fileSystem FileSystem
+}
+
+// NewValidator returns a new resource blob validator.
+func NewValidator(logger logger.Logger, fileSystem FileSystem) *Validator {
+	return &Validator{
+		logger:     logger,
+		fileSystem: fileSystem,
+	}
+}
+
+// Validate takes a ReadCloser containing a resource blob and checks that
+// the size and hash match the expected values. It downloads the blob to a
+// temporary file and returns a ReadCloser that deletes the temporary file on
+// closure.
+//
+// Returns [ErrUnexpectedHash] if the hash of the downloaded resource does not
+// match the expected hash.
+// Returns [ErrUnexpectedSize] if the size of the downloaded resource does not
+// match the expected size.
+func (v *Validator) Validate(
+	reader io.ReadCloser,
+	expectedSHA384 string,
+	expectedSize int64,
+) (_ io.ReadCloser, err error) {
+	defer func() {
+		_ = reader.Close()
+	}()
+
+	tmpFile, err := v.fileSystem.CreateTemp("", "resource-")
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	defer func() {
+		// Always close the file, we no longer require to have it open for
+		// this purpose. Another process/method can take over the file.
+		err = tmpFile.Close()
+
+		// If the download was successful, we don't need to remove the file.
+		// It is the responsibility of the caller to remove the file.
+		if err == nil {
+			v.logger.Warningf("closing temporary file: %v", err)
+			return
+		}
+
+		// Remove the temporary file if the download failed. If we can't
+		// remove the file, log a warning, so the operator can clean it up
+		// manually.
+		removeErr := v.fileSystem.Remove(tmpFile.Name())
+		if removeErr != nil {
+			v.logger.Warningf("failed to remove temporary file %q: %v", tmpFile.Name(), removeErr)
+		}
+	}()
+
+	freeDiskSpace, err := v.getFreeSpaceForPath(tmpFile.Name())
+	if err != nil {
+		return nil, errors.Errorf("getting free disk space on controller: %w", err)
```

Comment:
For now, just let Juju crash. If we want this pattern to work, create a spec for it.

Reply:
Dropped the commit with all this in.

Reply:
@Aflynn50 Please start a spec for this work. It's something we should look at for both charms and resources.